src/lib/comms/sol-http-common.c

```diff
@@ -74,52 +130,645 @@ sol_http_param_free(struct sol_http_param *params)
         return;
     }
     sol_vector_clear(&params->params);
+    if (params->arena) {
+        sol_arena_del(params->arena);
+        params->arena = NULL;
+    }
+}
+
+SOL_API int
+sol_http_encode_slice(struct sol_buffer *buf, const struct sol_str_slice value)
+{
+    int r;
+    size_t i, last_append;
+    bool escaped;
+
+    SOL_NULL_CHECK(buf, -EINVAL);
+
+    escaped = false;
+    last_append = 0;
+    for (i = 0; i < value.len; i++) {
+        unsigned char c = value.data[i];
+        if (!isalnum(c) && c != '_' && c != '~' && c != '.' && c != '-') {
+            if (!escaped) {
+                sol_buffer_init(buf);
```

Comment:
you can init before the loop and avoid !escaped. Just assume it's not escaped if after the loop end you didn't append anything (ie: would append the whole slice as-is), then you change the flag and set the pointer.

See https://github.com/solettaproject/soletta/blob/master/src/lib/parsers/sol-json.c#L696